knimin/lib/format.py

```diff
@@ -0,0 +1,273 @@
+
+    Parameters
+    ----------
+    instrument_type : {miseq, hiseq}
+        The instrument type.
+    labadmin_id : int
+        The ID of the sequencing run as informed by labadmin
+    run_name : str
+        The name of the run.
+    assay : str
+        The assay instrument (e.g., Kapa Hyper Plus)
+    fwd_cycles : int
+        The number of forward cycles
+    rev_cycles : int
+        The number of reverse cycles
+    pi_name : str
+        The name of the principle investigator
+    pi_email : str
+        The email address of the principle investigator
+    contact_0_name : str
+        The name of an additional contact person
+    contact_0_email : str
+        The email of an additional contact person
+    run_type : {"Target Gene", "Shotgun"}
+        Which data sheet structure to use
+    sample_information : iterable of dict
+        The sample information to load into the sample sheet. See note below.
+    contact_1_name : str, optional
+        The name of an additional contact.
+    contact_1_email : str, optional
+        The email of an additional contact.
+    contact_2_name : str, optional
+        The name of an additional contact.
+    contact_2_email : str, optional
+        The email of an additional contact.
+
+    Sample Sheet Note
+    -----------------
+    If the instrument type is a MiSeq, any lane information per-sample will be
+    disregarded. If instrument type is a HiSeq, each sample must include a
+    "lane" key.
+
+    If the run type is Target Gene, then sample details are disregarded
+    with the exception of determining the lanes.
+
+    IF the run type is shotgun, then the following keys are required:
+        - sample-id
+        - i7-index-id
+        - i7-index
+        - i5-index-id
+        - i5-index
+
+    Raises
+    ------
+    ValueError
+        If a contact name is specified and an email is omitted, and vice versa
+    ValueError
+        If an unknown run type is specified
+    ValueError
+        If the number of cycles are <= 0
+    ValueError
+        If an unknown instrument type is specified.
+
+    Return
+    ------
+    str
+        The formatted sheet.
+    """
+    def validate_contact(name, email):
+        if name is not None and email is None:
+            raise ValueError("email is missing for %s" % name)
+        elif name is None and email is not None:
+            raise ValueError("name is missing for %s" % email)
+
+    c1name = contact_1_name if contact_1_name is not None else ''
+    c1email = contact_1_email if contact_1_email is not None else ''
+    c2name = contact_2_name if contact_2_name is not None else ''
+    c2email = contact_2_email if contact_2_email is not None else ''
+
+    validate_contact(pi_name, pi_email)
+    validate_contact(contact_0_name, contact_0_email)
+    validate_contact(contact_1_name, contact_1_email)
+    validate_contact(contact_2_name, contact_2_email)
+
+    if fwd_cycles <= 0 or not isinstance(fwd_cycles, int):
```

Comment:
Is `fwd_cycles` a value that the user can type in the interface? I wonder if it's necessary to validate it in the backend.